### PR TITLE
[AutoBuild] Support model selection when building agents

### PR DESCRIPTION
**Pull request by whiskyboy**
## Why are these changes needed?

In some cases we need to set different agent with different backbone models, such as when we want to group a set of small LLMs with different capacities to complete a complex task.

This PR will allow `AgentBuilder()` accept a list of models as parameter to `agent_model`, and the `builder_manager` will select the most suitable model for each agent based on the agent/model profile and the task description.

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/contrib/agent_builder.py [hunk 1]
```diff
@@ -519,6 +568,17 @@ def build_from_library(
...
+        agent_model_config_list = autogen.config_list_from_json(
+            self.config_file_or_env,
+            file_location=self.config_file_location,
+            filter_dict={"model": self.agent_model},
+        )
+        if len(agent_model_config_list) == 0:
+            raise RuntimeError(
+                f"Fail to initialize agent model: {self.agent_model} does not exist in {self.config_file_or_env}. "
+                f'If you want to change this model list, please specify the "agent_model" in the constructor.'
+            )
```
Comment (thinkall) [inline autogen/agentchat/contrib/agent_builder.py:580]: Could we extract this part and  line 559-568 into a config checking function to  reduce duplication in `build` and  `build_from_library`?

**Review by thinkall — COMMENTED** [review]
Thank you very much, @whiskyboy.  LGTM!
I've only 2 tiny questions, but it's OK to ignore them.

File: notebook/autobuild_basic.ipynb [hunk 2]
```diff
@@ -1367,15 +1367,163 @@
...
+    "# Add a new model that is good at mathematical problem solving.\n",
+    "config_list.extend(\n",
+    "    [\n",
+    "        {\n",
+    "            # Choose your model name.\n",
+    "            \"model\": \"microsoft/phi-2\",\n",
+    "            \"api_key\": \"NotRequired\",\n",
+    "            # You need to provide a short description of your model here.\n",
+    "            \"profile\": \"A pre-trained model that is good at sovling mathematical reasoning problems.\",\n",
+    "        }\n",
```
Comment (LinxinS97) [inline notebook/autobuild_basic.ipynb:1402]: A description alone is not enough. Context length, cost, required GPU resources, and benchmark ranking are also necessary for model selection.

File: autogen/agentchat/contrib/agent_builder.py [hunk 3]
```diff
@@ -237,6 +255,12 @@ def _create_agent(
                     if self._is_port_open(self.host, port):
                         break
 
+                # Check if vLLM is installed.
+                try:
+                    import vllm  # noqa: F401
+                except ImportError:
+                    raise ImportError("vLLM is not installed. Please install vLLM by running 'pip install vllm'.")
```
[the remark this answers is not in the document]
Comment (whiskyboy) [reply autogen/agentchat/contrib/agent_builder.py:262]: Actually vLLM is not a hard dependency of autobuild, we only need it when we want to call a non-openai model locally: https://microsoft.github.io/autogen/docs/topics/non-openai-models/local-vllm

Without this check, the popen will hang there when vLLM is missing in the runtime.

File: notebook/autobuild_basic.ipynb (hunk 2 above)
Comment (LinxinS97) [reply notebook/autobuild_basic.ipynb:1402]: Yes, but context length and required GPU resources are crucial for successfully solving a task. We should add some restrictions or guidance for users to adopt this feature correctly, like splitting these fields out.

Comment (whiskyboy) [reply notebook/autobuild_basic.ipynb:1402]: Yeah, it makes sense to me. How do you prefer to use these crucial informations? Like longer context model has a higher priority, and check the gpu's avaliablity in runtime?

Comment (LinxinS97) [reply notebook/autobuild_basic.ipynb:1402]: > longer context model has a higher priority, and check the gpu's availability in runtime

LGTM!

File: autogen/agentchat/contrib/agent_builder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (whiskyboy) [reply autogen/agentchat/contrib/agent_builder.py]: Add vLLM and GPUtils to the extra dependency list

File: autogen/agentchat/contrib/agent_builder.py (hunk 1 above)
Comment (whiskyboy) [reply autogen/agentchat/contrib/agent_builder.py]: Done

File: notebook/autobuild_basic.ipynb (hunk 2 above)
Comment (whiskyboy) [reply notebook/autobuild_basic.ipynb]: @LinxinS97 Mind reviewing again?

**Comment by LinxinS97** [comment]
@whiskyboy Have you joined the discord? Could we have a meeting? I would like to share with you some latest information on autobuild and hear about your next plan.

**Comment by whiskyboy** [comment]
> @whiskyboy Have you joined the discord? Could we have a meeting? I would like to share with you some latest information on autobuild and hear about your next plan.

Sure, would love to. My discord name is whiskyboy0077.

BTW, are you an employee at Microsoft? If yes, can we talk via teams?

**Pull request closed without merging**